source/extensions/filters/http/file_system_buffer/filter.cc

```diff
@@ -389,22 +389,30 @@ bool FileSystemBufferFilter::maybeStorage(BufferedStreamState& state,
   // If there's more data in memory than there should be, put some into storage.
   if (state.memory_used_ > state.config_->memoryBufferBytesLimit() &&
       state.storage_used_ < state.config_->storageBufferBytesLimit()) {
-    auto dispatch = getSafeDispatch();
     if (!state.async_file_handle_) {
       // File isn't open yet - open it and then check again if we still need to store data.
       ENVOY_STREAM_LOG(debug, "memory buffer exceeded - creating buffer file", callbacks);
+      // We can't use getSafeDispatcher here because we need to close the file if the filter
+      // was deleted before the callback, not just do nothing.
       cancel_in_flight_async_action_ = config_->asyncFileManager().createAnonymousFile(
-          config_->storageBufferPath(),
-          [this, &state, dispatch](absl::StatusOr<AsyncFileHandle> file_handle) {
-            if (!file_handle.ok()) {
-              dispatch([this, status = file_handle.status()]() {
+          config_->storageBufferPath(), [this, me = std::weak_ptr(shared_from_this()),
+                                         dispatcher = &request_callbacks_->dispatcher(),
+                                         &state](absl::StatusOr<AsyncFileHandle> file_handle) {
+            dispatcher->post([this, me = std::move(me), &state, file_handle]() {
+              if (!me.lock()) {
+                // If we opened a file but the filter went away in the meantime, close the file
+                // to avoid leaving a dangling file handle.
+                if (file_handle.ok()) {
+                  file_handle.value()->close([](absl::Status) {}).IgnoreError();
+                }
```

Comment:
Quick question: Is there a reason we can't just use RAII to make sure that a file handle is closed if we end up destroying it without explicitly calling closed?

This seems like a safer way to make sure we don't have any fd leaks?  wdyt?

/wait-any

Reply:
Doing it that way would make it get closed in the envoy thread that we're trying not to block. I don't know if close can cause a long block, but open certainly can, so I wouldn't bet on 'no'.

Reply:
(Or, if you mean using RAII of the filter to call this close function, that doesn't work because the filter is destroyed before it knows whether it has a file handle.)

Reply:
OK sounds good, thanks.